### PR TITLE
Keep the unhandled-rejection queue append-only and track reported promises in a weak set

### PR DESCRIPTION
**Pull request by robobun**
### Problem

- N promises that reject and get a handler in the same turn cost O(N^2). `Promise.allSettled` over 100,000 async calls that throw takes 24 s on main. Node takes 0.5 s.
- The cause is `promiseRejectionTracker` (`src/jsc/bindings/ZigGlobalObject.cpp:1190`). Each `Handle` searched the queue with `Vector::removeFirstMatching`, then shifted every entry behind the hit. It also walked the batch that `handleRejectedPromises()` reports.

### Fix

- `Handle` only needs to know if the promise was reported. `handleRejectedPromises()` adds a promise to a `JSWeakSet` right before it reports it, and `Handle` asks that set. Nothing searches the queue. The design is @alii's, from review.
- The queue is append-only, in `src/jsc/bindings/RejectedPromiseQueue.{h,cpp}`. `Handle` pops the promise if it is the newest entry, with the handled entries before it. A full vector drops its handled entries before it grows.
- Verified: `test/js/node/process/process.test.js` (4 timing tests fail on main), and a differential fuzz with identical event logs.

### Background

- JSC calls `promiseRejectionTracker` with `Reject` when a promise rejects with no handler, and with `Handle` when it gets one. Bun reports what is left after the microtask drain. A handler after the report gets `rejectionHandled`.
- A `JSWeakSet` is the GC-managed set behind `WeakSet`. The GC removes a promise from it when the promise dies.

### Downsides

- A handled promise that is not the newest entry stays queued until it is at the end, the vector fills, or the drain. A loop that awaits each rejection one step late keeps up to 60 alive. Main keeps 2.
- While a reported promise is alive, each `Handle` of a promise that is not the newest entry pays a weak-set lookup: 67.5 ns against 64.8 ns per promise in batches of 64 (main: 82.8).

<details><summary>Notes</summary>

`append()` and `didHandle()` are `ALWAYS_INLINE` in the header: main inlines the old pair into the tracker, and the file split had stopped that (the +84 instructions per caught async throw of the cost report).

Release builds of the same commit (main 6d504dd983), linux x64, one pinned core, concurrent JIT off, min of 5 interleaved runs. "previous head" is e457ee91a7 of this PR (holes, the next-in-order check, a scan of 256, a `JSWeakMap` index). Nanoseconds per promise for `Promise.reject(i)` then `.catch(noop)`, K rejections per turn:

| shape | main | previous head | this PR |
| --- | --- | --- | --- |
| K = 1 | 148.9 | 148.2 | 147.9 to 150.4 |
| oldest first, K = 17 | 84.5 | 84.9 | 69.0 |
| oldest first, K = 64 | 83.2 | 83.2 | 64.4 |
| oldest first, K = 300 | 92.6 | 81.6 | 64.4 |
| oldest first, K = 1,000 | 1,781 | 82.5 | 64.8 |
| newest first, K = 300 | 122.8 | 79.3 | 79.6 |
| random, K = 64 | 84.1 | 87.9 | 64.9 |
| random, K = 256 | 107.4 | 115.6 | 64.9 |
| random, K = 512 | 138.8 | 170.1 | 64.3 |
| random, K = 1,000 | 596.0 | 163.6 | 64.8 |
| oldest first behind one unhandled rejection, K = 300 | 93.2 | 81.6 | 63.6 |
| `allSettled` loop, N = 100, reused Error | 310.2 | 307.9 | 296.5 |
| `allSettled` loop, N = 1,000, reused Error | 2,017 | 302.5 | 291.5 |
| `try { await f() } catch {}`, reused Error | 552.8 | 554.9 | 551.9 |
| `try { await f() } catch {}`, new Error | 809.6 | 810.8 | 813.7 |

K = 1 measured 147.9 without the full-vector check and 149.5 to 150.4 with a first version of it. The check is now the vector's own capacity check (`makeRoom()` runs where `Vector::append` would grow), and a later run measured main 160.4, previous head 158.6, this PR 159.3 on a busier host. The `try` rows are equal within noise.

Large N on this PR: `allSettled` over 100,000 throwing calls takes 148 ms (main 24,009, node 475). Attaching `.catch` to 160,000 rejected promises takes 24 ms oldest first, 36 ms in random order, 25 ms newest first (previous head 27 / 69 / 27, main 83 s / 49 s / 6.5 s as reported). An `unhandledRejection` listener that calls `promise.catch()` on 160,000 rejections takes 88 ms (previous head 99 ms, 12.7 s before).

Memory, the row that decided the full-vector sweep. One turn, 1,000,000 steps, each rejection (a new `Error`) awaited one step late:

| build | time | peak RSS growth |
| --- | --- | --- |
| main | 963 ms | 60 MB |
| previous head | 957 ms | 61 MB |
| append-only, handled entries leave only from the end | 3,360 ms | 1,165 MB |
| this PR (handled entries also leave when the vector is full) | 1,014 ms | 62 MB |

The sweep runs only where the vector would otherwise grow, from a capacity of 64 up, and the vector doubles when the sweep freed less than half. So `append()` stays amortized O(1), also for a burst of rejections that nobody handles.

Behaviour: a differential fuzz (random interleavings of rejections and handles in one turn, oldest-first, newest-first and mixed styles with strays, batches of 1 to 900, a listener that handles promises later in the batch or the one being reported, late handles, forced GC) gives byte-identical `unhandledRejection` / `rejectionHandled` logs on main and on this PR: 24 seeds on release builds, 8 more plus 3 with forced GC on the debug ASAN and UBSan build, which is clean. The behaviour test of the previous head passed unchanged on the new internals before I extended it.

Tests: the behaviour test pins the event order for rejections handled one step late, oldest first with stragglers, newest first, a few in no order, a long queue that shrinks and grows again, handles from the listener of the batch that is being reported (later promises get neither event, an already reported one gets `rejectionHandled`), and late handles while other rejections wait. Two GC checks, both through `heapStats().objectTypeCounts.Date` after `Bun.gc(true)`: the 143 reasons that only the queue can reach must all be alive (with a `visit()` that stops after 10 entries the child crashes, on debug with `ASSERTION FAILED: decontaminate()`, on release with a crash report), and a loop that awaits each rejection one step late must not keep more than 200 of 2,000 handled reasons alive (2,001 without the full-vector sweep, 47 to 60 with it, 2 on main). Five timing tests: oldest first, no particular order and from the listener at N = 100,000 (`ms < max(limit, 20 * baseline)` against the same calls on handled promises), and two ratio tests (batches of 17, 64 and 300 oldest first, and the `allSettled` loop at 100 and 1,000). On a release build of main four of them fail (killed at 20 s, 14,248 ms, 4,236 ms, and a ratio of 5.6).

One behaviour is different in a case the fuzz does not reach: a promise that is rejected in one global object and handled from another. Main asks the queue of the handler's global, does not find the promise, and emits `rejectionHandled` from there, also for a promise that was never reported. Now that global emits it only for a promise it reported itself. `node:vm` contexts are not affected: they send both operations to the main global.

Bun's semantics are unchanged where they differ from Node: a listener that handles a promise later in the batch suppresses both events for it (pinned by the existing test from #32554). Node v26 still reports it as unhandled.

History of this PR: the first version removed the search with holes and a hash index. A cost report showed 36 to 48 % more per promise in batches of 17 to 256, which led to the next-in-order check and inline fast paths. @Jarred-Sumner asked for a GC map instead of raw cell keys, and for a separate file. @alii then proposed this design, which needs neither the index nor the scan.

Found on the way and reported separately: `abort()` on a signal with N `{ once: true }` listeners is O(N^2) in `EventTarget` (80,000 listeners: 9.1 s, node 44 ms). It is not part of this change.

Other suites run on the debug build: all of `test/js/node/process/process.test.js`, the 20 node `parallel` tests whose names contain `unhandled` or `rejection`, `test/js/bun/module-graph/module-graph.test.ts`, `module-graph-gc.test.ts`.

</details>

---

**no test proof** · iteration 1 · platform-specific test(s) that do not run on this machine, deferring to CI, which covers all platforms: test/js/node/process/process.test.js

### CODE REVIEW

**Comment by robobun** [comment]
<div><sup>Updated 6:05 PM PT - Sep 17th, 2026</sup></div>

@robobun, your commit 2053f14a744934aed1281f4bdde17925a7105c12 is building: [`#117432`](https://buildkite.com/bun/bun/builds/117432)

**Comment by robobun** [comment]
Status for #43208.

Reproduced on release builds of main (8f8e16c577) and on 1.4.2, linux x64, with the program from the report:

```js
// settled.mjs <N>
const N = +process.argv[2] || 100000;
async function validate(item) { if (item.id >= 0) throw new Error("invalid item " + item.id); return item; }
const items = Array.from({ length: N }, (_, id) => ({ id }));
const t0 = performance.now();
const results = await Promise.allSettled(items.map(validate));
console.log(`${N} items: ${results.filter(r => r.status === "rejected").length} rejected, allSettled took ${(performance.now() - t0).toFixed(0)} ms`);
```

- main: 245 / 4,072 / 24,009 ms at N = 10,000 / 40,000 / 100,000. This PR: 13 / 63 / 141 ms. Node v26.3.0: 52 / 196 / 475 ms.
- The second walk (the batch that is being reported): register `process.on("unhandledRejection", (reason, promise) => promise.catch(() => {}))`, then reject 160,000 promises. 12.7 s before, 0.16 s with this PR.
- `test/js/node/process/process.test.js`: the two "handling N rejected promises is O(N)" tests fail on a release build without the `src/` change (the first run is killed at 20 s, the second measures 6,700 ms against a 1,000 ms limit). They pass with it on debug and release builds.

File: test/js/node/process/process.test.js [hunk 1]
```diff
@@ -1812,6 +1902,88 @@ it("process.hasUncaughtExceptionCaptureCallback", () => {
...
+      const promises = [];
+      for (let i = 0; i < ${N}; i++) promises.push(Promise.reject(i));
+      const start = performance.now();
+      for (const promise of promises) promise.catch(noop);
+      const ms = performance.now() - start;
+      for (let i = 0; i < 3; i++) await new Promise(r => setImmediate(r));
+      console.log(JSON.stringify({ ms, ...events }));
+    `);
+      expect(events).toEqual({ unhandledRejection: 0, rejectionHandled: 0 });
+      expect(ms).toBeLessThan(limit);
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/node/process/process.test.js]: Changed in e8eab3f897. The two tests now calibrate against linear work in the same child process: after the measured loop, the child makes the same `.catch()` calls again. The promises are handled by then, so none of those calls reaches the rejection tracker. The assertion is `ms < Math.max(limit, 20 * baseline)`.

- A runner that is too slow for the fixed limit still passes while the measured calls stay within 20x of that baseline. Measured `ms / baseline`: 0.7 to 2.2 with the fix (debug and release), 190 to 880 without it at N = 30,000, and more at N = 100,000.
- I did not use N against 4N. With the fix the release regions take 20 to 30 ms, so one GC pause in the larger run moves that ratio more than the complexity does. The fixed limit absorbs a pause, the baseline absorbs a slow machine, and the test fails only when both say quadratic.
- The per-test timeout stays. A debug build needs 3 to 4 s for one of these child processes, and the default is 5 s. The 20 s spawn timeout makes a quadratic run an assertion failure (`signalCode: "SIGKILL"`), not a hang. In CI the explicit 30 s is lower than the runner's 90 s default. `setTimeout.test.js` ("GC of many id-accessed timers is not quadratic") and `url.test.ts` do the same for the same kind of test.

File: src/jsc/bindings/ZigGlobalObject.cpp [hunk 2]
```diff
@@ -3340,9 +3338,20 @@ RefPtr<Performance> GlobalObject::performance()
...
 extern "C" void Bun__handleRejectedPromise(Zig::GlobalObject* JSGlobalObject, JSC::JSPromise* promise, JSC::EncodedJSValue rejectionOwner);
 
+// Up to this many promises are cheaper to scan than to hash.
+static constexpr unsigned rejectedPromiseScanLimit = 16;
+
 void GlobalObject::RejectedPromiseQueue::append(JSC::VM& vm, JSC::JSCell* owner, JSC::JSPromise* promise, JSC::JSObject* rejectionOwner)
 {
     WTF::Locker locker { owner->cellLock() };
+    // Full, and at least half of it is holes: squeeze them out instead of growing.
+    // Costs what the reallocation would have, so append() stays amortized O(1).
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/jsc/bindings/ZigGlobalObject.cpp]: Shortened to one line in bbcd3a0cf1.

File: src/jsc/bindings/ZigGlobalObject.cpp [hunk 3]
```diff
@@ -3351,21 +3360,71 @@ void GlobalObject::RejectedPromiseQueue::append(JSC::VM& vm, JSC::JSCell* owner,
 bool GlobalObject::RejectedPromiseQueue::remove(JSC::JSCell* owner, JSC::JSPromise* promise)
 {
     WTF::Locker locker { owner->cellLock() };
-    return m_entries.removeFirstMatching([&](Entry& entry) { return entry.promise.get() == promise; });
+    // Newest first: the newest rejection is the usual one to be handled. A long
+    // queue is scanned no further than that entry.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/jsc/bindings/ZigGlobalObject.cpp]: Shortened to one line in bbcd3a0cf1.

File: src/jsc/bindings/ZigGlobalObject.cpp [hunk 4]
```diff
@@ -3351,21 +3360,71 @@ void GlobalObject::RejectedPromiseQueue::append(JSC::VM& vm, JSC::JSCell* owner,
...
+    }
+
+    if (index + 1 < size) {
+        m_entries[index].promise.clear();
+        m_entries[index].rejectionOwner.clear();
+        ++m_holes;
+        return true;
+    }
+    // The last entry is never a hole, so a loop that rejects and then handles
+    // one promise at a time keeps reusing the same slot.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/jsc/bindings/ZigGlobalObject.cpp]: Shortened to one line in bbcd3a0cf1.

File: src/jsc/bindings/ZigGlobalObject.h [hunk 5]
```diff
@@ -848,7 +848,15 @@ class GlobalObject : public Bun::GlobalScope {
             JSC::WriteBarrier<JSC::Unknown> promise; // JSPromise
             JSC::WriteBarrier<JSC::Unknown> rejectionOwner; // JSModuleGraph or null
         };
+        // In rejection order. remove() leaves a hole (an empty promise) in place of
+        // any entry but the last, so the entries after it keep their index.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/jsc/bindings/ZigGlobalObject.h]: Shortened to one line in bbcd3a0cf1.

File: src/jsc/bindings/ZigGlobalObject.h [hunk 6]
```diff
@@ -848,7 +848,15 @@ class GlobalObject : public Bun::GlobalScope {
...
         WTF::Vector<Entry> m_entries;
+        unsigned m_holes { 0 };
+        // promise -> index, for m_entries[0..m_indexed). remove() extends it only when
+        // it has to search a long queue, so handling the newest rejection, or one of
+        // a few, does not build it.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/jsc/bindings/ZigGlobalObject.h]: Shortened to one line in bbcd3a0cf1.

File: test/js/node/process/process.test.js [hunk 7]
```diff
@@ -1787,6 +1787,96 @@ describe.concurrent(() => {
     expect({ stdout: stdout.trim(), stderr, exitCode }).toEqual({ stdout: "ok", stderr: "", exitCode: 0 });
   });
 
+  // Pins which events fire, and in what order, for every way the queue of
+  // unreported rejections finds and drops a handled promise: a hole left in the
+  // middle, compaction, a pop from the back, a short queue, and the batch that
+  // is being reported. The "handling N rejected promises is O(N)" tests below
+  // cover the cost of that lookup.
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/node/process/process.test.js]: Added in f475ca20d0 as the `regrown` section: 20 rejections, a handler on one in the middle (this builds the index), a handler on the newest (the pop from the back), two more rejections, then a handler on the older one and on the newer one. It expects no `rejectionHandled` and neither new promise in `seen`.

I checked it against the mutation. With the `m_indexed` clamp deleted, the section reports `rejectionHandled: 1` and the test fails. With the clamp it passes on the debug build, and on a release build of main (the old implementation), so the expectation matches the old semantics.

File: test/js/node/process/process.test.js [hunk 8]
```diff
@@ -1812,6 +1902,99 @@ it("process.hasUncaughtExceptionCaptureCallback", () => {
...
+  //   after:  release 25 / 20 ms, debug+ASAN 700 / 440 ms
+  const limit = isDebug || isASAN ? 5_000 : 1_000;
+  // A machine too slow for `limit` still passes if the calls stay within 20x of
+  // `baseline`: the same calls again, on promises that are handled by then, so
+  // none of them reaches the rejection tracker. Measured ms / baseline: 0.7 to
+  // 2.2 after, 600 and up before.
+  const expectLinear = ({ ms, baseline }) => expect(ms).toBeLessThan(Math.max(limit, 20 * baseline));
+  // A debug build needs about 2 s to start and make the promises, too close to
+  // the 5 s default. The spawn timeout turns a quadratic run into a failure.
+  const timeout = 30_000;
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/node/process/process.test.js:1935]: I kept N = 100,000 and the timeout. A smaller N does not separate the two builds in a debug binary, and the debug binary is what a local fail-before run uses.

The quadratic term needs a large N to dominate in debug, because every `.catch()` call costs 5 to 10 us there. Measured `ms / baseline` without the fix on a debug+ASAN build:

| N | first test | second test |
| --- | --- | --- |
| 25,000 | about 15x | about 41x |
| 50,000 | about 29x | about 84x |
| 100,000 | about 58x | more (killed at 20 s) |

At 25,000 the first test would pass without the fix under the 20x bound. A lower bound does not help: with the fix the release ratio reaches 2.2x, and at a small N the measured region is a few milliseconds, so one GC pause moves the ratio more than the complexity does. The fixed `limit` would also have to shrink with N, or it hides the quadratic run on a release build (the second test takes 400 ms at N = 30,000 without the fix).

At 100,000 every combination separates with margin: with the fix 0.7x to 2.2x, without it 58x and more on debug and 350x and more on release.

A debug child then needs 3 to 4 s, mostly startup and making the promises, and the default is 5 s. That is the only reason for the explicit 30 s. In CI it lowers the runner's 90 s default. A release child takes about 0.15 s. `setTimeout.test.js` ("GC of many id-accessed timers is not quadratic") and `url.test.ts` carry the same kind of timeout for the same kind of test.

**Comment by robobun** [comment]
I worked on the same report in parallel and stopped when I saw this PR. My branch is [`robobun/6e5d982f/rejected-promise-queue-hash`](https://github.com/oven-sh/bun/compare/main...robobun/6e5d982f/rejected-promise-queue-hash) (one commit, 92b6c84043, no PR). The design is the same as here: entries are nulled in place, and a promise-to-index map is built on demand. Two differences showed up in measurements. Both can be folded into this PR.

**1. `append()` and `remove()` are no longer inlined into `promiseRejectionTracker`.** With the map code inside them, clang keeps them out of line. Instructions executed inside `promiseRejectionTracker` per promise (the `Reject` call plus the `Handle` call), counted with gdb (`stepi` from the entry of the function to its return address). Release builds of b4315fa551 with only `ZigGlobalObject.{cpp,h}` changed:

| shape | main | this PR | branch |
| --- | --- | --- | --- |
| reject one, handle it, repeat | 139 | 223 | 145 |
| 10 per turn, handled oldest first | 159 | 275 | 174 |
| 10 per turn, newest first | 162 | 229 | 150 |
| 100 per turn, oldest first | 245 | 623 | 177 |
| 100 per turn, newest first | 387 | 230 | 150 |

The branch marks `append()`, `remove()` and the removal itself `ALWAYS_INLINE`, and moves the scan, the map and the compaction into `NEVER_INLINE` functions. Before that split the first row measured 211 on the branch too.

**2. Oldest first always goes through the map.** That is the order in which `Promise.allSettled` attaches handlers. The branch keeps the index of the first entry that is not null (`m_begin`). It checks the last entry and that entry before anything else, so newest first and oldest first never hash. Nanoseconds per promise, best of 9 interleaved runs, promises rejected in one turn and handled one turn later, oldest first:

| per turn | main | this PR | branch |
| --- | --- | --- | --- |
| 1 | 230 | 231 | 231 |
| 10 | 105 | 110 | 110 |
| 100 | 102 | 178 | 98 |
| 1,000 | 1,847 | 145 | 94 |

`await Promise.allSettled()` over N async calls that throw, in a loop, ns per promise (main / this PR / branch): N = 100: 591 / 657 / 575. N = 1,000: 2,493 / 766 / 707. One call with N = 160,000: 64.9 s / 194 ms / 159 ms.

Other notes:

- The scan limit. In instruction counts the map breaks even with a scan between 64 and 128 entries. In wall time the break-even is near 300. The branch uses 64.
- On this machine (Xeon 8375C, glibc 2.41) main is 5.5x slower than a build from before #42590 (c6b7fcb5bb) at 1,000 rejections per turn, not 2x (2,100 ms against 390 ms CPU per million). 1,000 entries of 16 bytes are 16 KB, and glibc does an overlapping memmove of that size with `rep movsb`. With a 16-byte shift that takes about 8 µs. The same move of 8 KB takes about 100 ns.
- The tests of the branch are in their own file, `test/js/node/process/process-unhandled-rejections.test.ts`, which is kept out of the parallel batch because it compares timings. They are a model-based check of both events, and a timing test with three handling orders (oldest first, newest first, scattered).
- This PR also makes the lookup in the batch that is being reported constant time. The branch does not have that.

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
Put the RejectedPromiseQueue code in a separate file.

**Comment by robobun** [comment]
Done in e3eaf33a99. `RejectedPromiseQueue` and `InFlightRejections` (the drained batch with its `tailContains()` lookup) are now in `src/jsc/bindings/RejectedPromiseQueue.h` and `RejectedPromiseQueue.cpp`, in namespace `Bun`.

- `ZigGlobalObject.h` includes the new header and keeps the two members with their existing comments. `ZigGlobalObject.cpp` keeps `promiseRejectionTracker` and `handleRejectedPromises()`, and no longer has any queue code.
- The code did not change in the move. `visit()` has explicit instantiations for `JSC::AbstractSlotVisitor` and `JSC::SlotVisitor`, like `BunCommonStrings.cpp`.
- The new `.cpp` compiles on its own (checked with `-fsyntax-only` from its `compile_commands.json` entry) and inside its unified bundle. The rejection tests pass on the debug build.

**Comment by alii** [comment]
@robobun Is this a better fix? https://github.com/oven-sh/bun/commit/92b6c84043d54043f5665bfd1a3deddc16233fb9

**Review by Jarred-Sumner — DISMISSED** [review]
the hashmap leaves open a UAF if it gets GC'd in-between. maybe it should use one of the GC maps?

**Comment by alii** [comment]
@robobun about my earlier question: yes, please fold the two measured parts of 92b6c84043d into this PR before it merges.

1. Add the front index (m_begin) so oldest-first and newest-first handling never touches the hash.
2. Put the append and remove fast paths inline in RejectedPromiseQueue.h. Keep the map, scan and compaction out of line. Raise the scan limit toward 64 if your numbers agree.
3. Keep the in-flight batch fix from this head. The branch does not have it.
4. Add timing cases to process.test.js: batches of 17 to 300 handled oldest first, and the allSettled loop at N=100 and N=1,000. Re-measure against main for those shapes.
5. Add a GC test: call Bun.gc(true) before drain() in oneStepLate, with the promises reachable only through the queue.

**Comment by robobun** [comment]
@Jarred-Sumner @alii Understood, I will do both.

- **GC map.** Today every key of `m_indices` has a live, visited entry in `m_entries`, and every key of the in-flight map is in the `MarkedArgumentBuffer`, so no key can die while it is in a map. That is an invariant kept by hand, not by construction. I will replace both maps with a lazily created `JSWeakMap` (promise to index), created outside the cell lock. With the change below the map is only used for out-of-order handles in a queue longer than the scan limit, so its cost is off the common paths.
- **92b6c84043.** Yes, its two measured parts are the right ones, and the pushed head lacks both. I already have them locally after the cost report on this PR: inline `append()` / `remove()` fast paths in `RejectedPromiseQueue.h` (the file split had stopped the inlining, that was the +84 instructions), and O(1) checks of the newest entry and of the next entry in order, so oldest-first and newest-first batches never build the map. With that, batches of 1 to 256 measure at parity with main in all three orders and the `allSettled` cells of the cost report are at parity too.
- I keep the in-flight batch fix, add the timing cases (batches of 17 to 300 oldest first, the `allSettled` loop at N = 100 and N = 1,000) and the `Bun.gc(true)` case in `oneStepLate`.
- Scan limit: I will post the numbers for 16, 64 and 256. So far a scan beats the map up to about 256 entries for out-of-order handles, once a scanned queue is compacted at half holes.

I am building 92b6c84043 on current main now to measure it on the same cells, and I will post the table with the push.

**Comment by robobun** [comment]
Pushed as a5dd7e9080 and e457ee91a7 (the branch is also rebased on main 6d504dd983).

**@Jarred-Sumner, the map.** Both maps are a `JSWeakMap` now (promise to index), so no raw cell pointer is a key. The map is made on the first lookup that needs it, outside the cell lock, because it is a GC allocation. `find()` also checks that the entry it is pointed at holds the promise. The JS thread now reads the queue without the lock (only it writes), so the lock covers the writes only, and a scan or an index build no longer blocks the GC thread. The behaviour test runs `Bun.gc(true)` while the index exists and while the map of the reported batch exists.

**@alii, your five points.**

1. Oldest-first and newest-first handles never reach the map. `remove()` checks the newest entry, then the entry after its last hit. I kept that form and not a plain front index because of one shape: with one older rejection still unhandled, a front index points at it for the whole turn, and every oldest-first handle behind it falls to the scan or the map (rows 5 and 6 below). For a batch from an empty queue the two are the same entry.
2. `append()` and `remove()` are `ALWAYS_INLINE` in `RejectedPromiseQueue.h`. `find()` (scan and map) and `compact()` are out of line in the `.cpp`. The symbols of the first two are gone from the binary again, as on main. Scan limit: 256. With the weak map a scan wins up to there (random order, K = 256: 115 ns scanned, 169 ns through the map). A scanned queue is compacted at half holes so that the scan stays as long as the live entries. With a `HashMap` the crossover was near 128 to 192.
3. The in-flight batch fix stays, with a `JSWeakMap` too. 92b6c84 on the same main takes 695 ms for a listener that handles 40,000 rejections, this PR takes 112 ms for 160,000.
4. New timing cases in `process.test.js`: batches of 17, 64 and 300 oldest first, and the `allSettled` loop at 100 and 1,000 (a promise may not cost 3 times more in the larger shape), plus handles in no particular order at 100,000. On a release build of main the `allSettled` ratio is 5.6 and the three N = 100,000 tests fail. The batch case passes on main, which has no cliff below about 500.
5. `oneStepLate` calls `Bun.gc(true)` before `drain()`. The promises that stay unhandled there are reachable only through the queue.

Nanoseconds per promise, release builds of the same main, one pinned core, concurrent JIT off, min of 5 interleaved runs. "92b6c84" is that commit's `src/` rebuilt on the same main:

| shape | main | 92b6c84 | this PR |
| --- | --- | --- | --- |
| K = 1 | 152.0 | 151.9 | 150.1 |
| oldest first, K = 17 / 64 / 300 | 86.9 / 86.6 / 94.6 | 87.2 / 84.1 / 80.2 | 87.9 / 86.5 / 82.7 |
| oldest first, K = 1,000 | 1,783 | 81.2 | 83.0 |
| newest first, K = 64 / 300 | 84.3 / 122.7 | 82.3 / 79.9 | 81.3 / 79.5 |
| oldest first behind one unhandled rejection, K = 64 | 83.3 | 118.5 | 83.0 |
| oldest first behind one unhandled rejection, K = 300 | 92.9 | 116.0 | 81.3 |
| random order, K = 64 / 128 / 256 | 84.1 / 89.2 / 106.7 | 87.7 / 102.1 / 106.8 | 88.2 / 95.5 / 114.6 |
| random order, K = 512 / 1,000 | 139.8 / 594.4 | 106.0 / 117.0 | 169.3 / 167.4 |
| `allSettled` loop, N = 100 / 1,000, reused Error | 344.7 / 2,674 | 343.3 / 498.2 | 318.5 / 484.2 |
| `allSettled` loop, N = 100 / 1,000, new Error | 900.2 / 3,032 | 887.9 / 865.8 | 888.5 / 805.7 |
| `try { await f() } catch {}`, reused / new Error | 560.8 / 831.3 | 570.4 / 868.2 | 557.3 / 818.9 |

The one place where this PR is slower than main is random-order handling of about 300 to 700 rejections per turn (169 ns against 140 ns at 512). That is the price of the weak map against a raw `HashMap` (106 ns in 92b6c84). It is in the Downsides of the description. A differential fuzz still gives byte-identical `unhandledRejection` / `rejectionHandled` logs on main and on this head (24 seeds, 4 more with forced GC), and the debug ASAN build is clean on it.

**Review by alii — CHANGES_REQUESTED** [review]
Requesting changes. I think a smaller design does the same job, so please try it before this lands.

Keep the queue append-only. Reject appends. The drain loop already skips handled promises. Mark a promise reported (a JSWeakSet, so no WebKit bump) right before Bun__handleRejectedPromise. On Handle: if the promise is reported, call Bun__handleHandledPromise. Otherwise pop the tail if it is this promise and trim trailing handled entries.

@robobun please prototype this on this branch, then run the existing behavior test unchanged, the four timing tests, the differential fuzz (event logs must be byte identical) and the timing table in the body. If every row is at or below this PR, replace the queue internals and drop the index, scan, compaction and in-flight code. If a row is worse, say which one and we merge this as is.

Also, I think the GC section of the behavior test would not fail if visit() stopped at the first empty barrier, on release or debug builds. Please make it fail in that case.

**Comment by robobun** [comment]
@alii Done in d135ff38a0 and 1394800200. Your design replaces the queue internals. The index, the scan, the holes, the compaction and the in-flight code are gone.

**What I ran on the prototype, in your order**

- The behaviour test, unchanged: passes (debug and release).
- The timing tests (five now): pass on debug and release. On a release build of main four of them still fail.
- The differential fuzz: byte-identical `unhandledRejection` / `rejectionHandled` logs against main. 24 seeds on release builds, 8 more plus 3 with forced GC on the debug ASAN build.
- The timing table: every row is at or below the previous head. Most rows are well below, because a `Handle` of a promise that is not the newest entry now does nothing at all.

| ns per promise | main | previous head | this head |
| --- | --- | --- | --- |
| K = 1 | 148.9 | 148.2 | 147.9 to 150.4 |
| oldest first, K = 17 / 64 / 300 / 1,000 | 84.5 / 83.2 / 92.6 / 1,781 | 84.9 / 83.2 / 81.6 / 82.5 | 69.0 / 64.4 / 64.4 / 64.8 |
| newest first, K = 300 | 122.8 | 79.3 | 79.6 |
| random, K = 64 / 256 / 512 / 1,000 | 84.1 / 107.4 / 138.8 / 596.0 | 87.9 / 115.6 / 170.1 / 163.6 | 64.9 / 64.9 / 64.3 / 64.8 |
| oldest first behind one unhandled rejection, K = 300 | 93.2 | 81.6 | 63.6 |
| `allSettled` loop, N = 100 / 1,000 | 310.2 / 2,017 | 307.9 / 302.5 | 296.5 / 291.5 |
| `try { await f() } catch {}`, reused / new Error | 552.8 / 809.6 | 554.9 / 810.8 | 551.9 / 813.7 |

The random-order row where the previous head was slower than main (170 ns against 139 ns at K = 512) is now 64 ns.

**One row was worse, and it is not a timing row.** With handled entries leaving only from the end, a turn that never handles its newest rejection keeps every handled promise and its reason until the drain. One turn, 1,000,000 steps, each rejection (a new `Error`) awaited one step late:

| build | time | peak RSS growth |
| --- | --- | --- |
| main | 963 ms | 60 MB |
| previous head | 957 ms | 61 MB |
| the design exactly as you wrote it | 3,360 ms | 1,165 MB |
| this head | 1,014 ms | 62 MB |

I kept your design and added one thing for that row: when the vector is full, `append()` removes the handled entries before it grows, from a capacity of 64 up, and doubles when that freed less than half. It is the vector's own capacity check (`makeRoom()` runs where `Vector::append` would grow), so the common path has no extra branch. If you prefer to merge without it, say so, but I would not ship the 1 GB row.

**The GC section.** You were right, it could not fail. It now counts live objects: the reasons of the 143 promises that stay unhandled in `oneStepLate` are `Date`s that only the queue can reach, and after `Bun.gc(true)` `heapStats().objectTypeCounts.Date` must be 143. I checked it against a `visit()` that stops after 10 entries: the child dies on debug (`ASSERTION FAILED: decontaminate()`) and on release (crash report), so the test fails on both. A second phase pins the memory row above: a loop that awaits each rejection one step late may not keep more than 200 of 2,000 handled reasons alive (2,001 without the full-vector sweep, 47 to 60 with it, 2 on main).

**Two things to know.**

- While a reported promise is still alive, a `Handle` of a promise that is not the newest entry pays one weak-set lookup: 67.5 ns against 64.8 ns per promise in batches of 64 (main 82.8). With no reported promise alive the set is not touched.
- A promise rejected in one global object and handled from another: main asks the handler's global, does not find it, and emits `rejectionHandled` there, also for a promise that was never reported. Now a global emits it only for a promise it reported itself. `node:vm` contexts send both operations to the main global, so they are not affected.

The PR title and description are updated to this design.

**Comment by robobun** [comment]
The description was rewritten for this design right after the push of d135ff38a0 (the title too). The weak map, the 256-entry scan and the `m_indexed` clamp now appear only in the Notes, as the "previous head" column of the table and in the history of the PR.

**Review by alii — APPROVED** [review]
